File: python/tests/test_greedy.py
import unittest
import sys
import os

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

from greedy.activity_selection import activity_selection, activity_selection_with_names
from greedy.huffman import build_huffman_tree, build_codes, encode, decode, huffman_compress


# ── Activity Selection ───────────────────────────────────────────────────────

class TestActivitySelection(unittest.TestCase):

    def test_basic_case(self):
        acts = [(1,4), (3,5), (0,6), (5,7), (3,9), (5,9), (6,10), (8,11)]
        result = activity_selection(acts)
        self.assertEqual(len(result), 3)

    def test_no_overlap(self):
        # All activities are non-overlapping — should select all
        acts = [(0,1), (2,3), (4,5), (6,7)]
        self.assertEqual(len(activity_selection(acts)), 4)

    def test_all_overlap(self):
        # All overlap with each other — should select only 1
        acts = [(0,10), (1,9), (2,8), (3,7)]
        self.assertEqual(len(activity_selection(acts)), 1)

    def test_empty_input(self):
        self.assertEqual(activity_selection([]), [])

    def test_single_activity(self):
        self.assertEqual(activity_selection([(2, 5)]), [(2, 5)])

    def test_result_is_non_overlapping(self):
        acts = [(1,4), (3,5), (0,6), (5,7), (3,9), (5,9), (6,10), (8,11)]
        result = activity_selection(acts)
        for i in range(len(result) - 1):
            self.assertGreaterEqual(result[i + 1][0], result[i][1])

    def test_unsorted_input(self):
        # Should handle unsorted input correctly
        acts = [(8,11), (1,4), (5,7), (0,6)]
        result = activity_selection(acts)
        self.assertEqual(len(result), 3)

    def test_adjacent_activities(self):
        # Finish time of one == start time of next — should both be selected
        acts = [(0,2), (2,4), (4,6)]
        self.assertEqual(len(activity_selection(acts)), 3)

    def test_with_names(self):
        acts = [
            {'name': 'A', 'start': 1, 'finish': 4},
            {'name': 'B', 'start': 3, 'finish': 5},
            {'name': 'C', 'start': 5, 'finish': 7},
        ]
        result = activity_selection_with_names(acts)
        self.assertEqual(len(result), 2)
        self.assertEqual(result[0]['name'], 'A')
        self.assertEqual(result[1]['name'], 'C')


# ── Huffman Coding ───────────────────────────────────────────────────────────

class TestHuffman(unittest.TestCase):

    def test_codes_generated(self):
        codes = build_codes(build_huffman_tree("hello"))
        self.assertIn('h', codes)
        self.assertIn('e', codes)
        self.assertIn('l', codes)
        self.assertIn('o', codes)

    def test_frequent_char_shorter_code(self):
        # 'l' appears twice, others once — 'l' should have shorter or equal code
        codes = build_codes(build_huffman_tree("hello"))
        self.assertLessEqual(len(codes['l']), len(codes['h']))

    def test_encode_decode_roundtrip(self):
        text = "hello world"
        root = build_huffman_tree(text)
        codes = build_codes(root)
        encoded = encode(text, codes)
        decoded = decode(encoded, root)
        self.assertEqual(decoded, text)

    def test_encode_decode_long_text(self):
        text = "the quick brown fox jumps over the lazy dog"
        root = build_huffman_tree(text)
        codes = build_codes(root)
        self.assertEqual(decode(encode(text, codes), root), text)

    def test_single_char_text(self):
        text = "aaaa"
        result = huffman_compress(text)
        self.assertEqual(decode(result['encoded'], result['root']), text)

    def test_compression_reduces_size(self):
        # Repeated characters should compress well
        text = "aaaaaabbbbccdd"
        result = huffman_compress(text)
        self.assertLess(result['compressed_bits'], result['original_bits'])

    def test_all_codes_unique(self):
        codes = build_codes(build_huffman_tree("abcdefgh"))
        values = list(codes.values())
        self.assertEqual(len(values), len(set(values)))

    def test_codes_are_prefix_free(self):
        # No code should be a prefix of another (fundamental Huffman property)
        codes = build_codes(build_huffman_tree("abracadabra"))
        code_list = list(codes.values())
        for i, c1 in enumerate(code_list):
            for j, c2 in enumerate(code_list):
                if i != j:
                    self.assertFalse(
                        c2.startswith(c1),
                        f"Code '{c1}' is a prefix of '{c2}' — not prefix-free!"
                    )

    def test_empty_text(self):
        self.assertIsNone(build_huffman_tree(""))
        self.assertEqual(huffman_compress(""), {})

    def test_compression_stats_present(self):
        result = huffman_compress("hello world")
        self.assertIn('compression_pct', result)
        self.assertIn('original_bits', result)
        self.assertIn('compressed_bits', result)


if __name__ == "__main__":
    unittest.main()

File: java/tests/TestGreedy.java
package tests;

import org.junit.Test;
import static org.junit.Assert.*;
import java.util.*;
import greedy.ActivitySelection;
import greedy.ActivitySelection.Activity;
import greedy.HuffmanCoding;
import greedy.HuffmanCoding.Node;

/**
 * Test suite for Greedy Algorithms: Activity Selection and Huffman Coding.
 *
 * Compile:
 *   javac -cp ".;lib/junit-4.13.2.jar" -d out java/greedy/*.java java/tests/TestGreedy.java
 * Run:
 *   java -cp "out;lib/junit-4.13.2.jar;lib/hamcrest-core-1.3.jar" org.junit.runner.JUnitCore tests.TestGreedy
 */
public class TestGreedy {

    // ── Activity Selection ───────────────────────────────────────────────────

    private List<Activity> acts(int[][] data) {
        List<Activity> list = new ArrayList<>();
        for (int[] d : data) list.add(new Activity(d[0], d[1]));
        return list;
    }

    @Test
    public void activitySelection_basicCase() {
        List<Activity> result = ActivitySelection.select(
            acts(new int[][]{{1,4},{3,5},{0,6},{5,7},{3,9},{5,9},{6,10},{8,11}})
        );
        assertEquals(3, result.size());
    }

    @Test
    public void activitySelection_noOverlap() {
        assertEquals(4, ActivitySelection.select(
            acts(new int[][]{{0,1},{2,3},{4,5},{6,7}})
        ).size());
    }

    @Test
    public void activitySelection_allOverlap() {
        assertEquals(1, ActivitySelection.select(
            acts(new int[][]{{0,10},{1,9},{2,8},{3,7}})
        ).size());
    }

    @Test
    public void activitySelection_emptyInput() {
        assertTrue(ActivitySelection.select(new ArrayList<>()).isEmpty());
    }

    @Test
    public void activitySelection_singleActivity() {
        List<Activity> result = ActivitySelection.select(acts(new int[][]{{2,5}}));
        assertEquals(1, result.size());
    }

    @Test
    public void activitySelection_nonOverlapping() {
        List<Activity> result = ActivitySelection.select(
            acts(new int[][]{{1,4},{3,5},{0,6},{5,7},{3,9},{5,9},{6,10},{8,11}})
        );
        for (int i = 0; i < result.size() - 1; i++)
            assertTrue(result.get(i + 1).start >= result.get(i).finish);
    }

    @Test
    public void activitySelection_adjacentActivities() {
        assertEquals(3, ActivitySelection.select(
            acts(new int[][]{{0,2},{2,4},{4,6}})
        ).size());
    }

    // ── Huffman Coding ───────────────────────────────────────────────────────

    @Test
    public void huffman_codesGenerated() {
        Map<Character, String> codes = HuffmanCoding.buildCodes(HuffmanCoding.buildTree("hello"));
        assertTrue(codes.containsKey('h'));
        assertTrue(codes.containsKey('e'));
        assertTrue(codes.containsKey('l'));
        assertTrue(codes.containsKey('o'));
    }

    @Test
    public void huffman_frequentCharShorterCode() {
        Map<Character, String> codes = HuffmanCoding.buildCodes(HuffmanCoding.buildTree("hello"));
        assertTrue(codes.get('l').length() <= codes.get('h').length());
    }

    @Test
    public void huffman_encodeDecodeRoundtrip() {
        String text = "hello world";
        Node root = HuffmanCoding.buildTree(text);
        Map<Character, String> codes = HuffmanCoding.buildCodes(root);
        assertEquals(text, HuffmanCoding.decode(HuffmanCoding.encode(text, codes), root));
    }

    @Test
    public void huffman_longTextRoundtrip() {
        String text = "the quick brown fox jumps over the lazy dog";
        Node root = HuffmanCoding.buildTree(text);
        Map<Character, String> codes = HuffmanCoding.buildCodes(root);
        assertEquals(text, HuffmanCoding.decode(HuffmanCoding.encode(text, codes), root));
    }

    @Test
    public void huffman_allCodesUnique() {
        Map<Character, String> codes = HuffmanCoding.buildCodes(HuffmanCoding.buildTree("abcdefgh"));
        Set<String> values = new HashSet<>(codes.values());
        assertEquals(codes.size(), values.size());
    }

    @Test
    public void huffman_prefixFree() {
        Map<Character, String> codes = HuffmanCoding.buildCodes(HuffmanCoding.buildTree("abracadabra"));
        List<String> codeList = new ArrayList<>(codes.values());
        for (int i = 0; i < codeList.size(); i++)
            for (int j = 0; j < codeList.size(); j++)
                if (i != j)
                    assertFalse(codeList.get(j).startsWith(codeList.get(i)));
    }

    @Test
    public void huffman_compressionReducesSize() {
        int[] stats = HuffmanCoding.compressionStats("aaaaaabbbbccdd");
        assertTrue(stats[1] < stats[0]);
    }

    @Test
    public void huffman_emptyText() {
        assertNull(HuffmanCoding.buildTree(""));
    }
}
